sql: implement index changes of PARTITION ALL BY on CREATE TABLE

otan opened the pull request:
Resolves #58733 

Release note (sql change): Implemented PARTITION ALL BY syntax for CREATE
TABLE, which automatically partitions all indexes on the current table
with the same partitioning scheme.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58748)
<!-- Reviewable:end -->

ajstorm reviewed (changes requested):
Nit in commit comment above: "automatically partitions the table, and all indexes, with the same partitioning scheme."

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58748#-:-MQr89eYCB-ntVr9OcRe:b3bgnrr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit, line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58748#-MQrL-vO1qQ3eNRz3UHk:-MQrL-vO1qQ3eNRz3UHl:b-cd9ith) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b8309bb678afabc85d11022c7a0e8cdbce6f0b9/pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit#L181)):*
> ```
> 
> query T
> SELECT create_statement FROM [SHOW CREATE TABLE t]
> ```

I might be missing something, but will this `SHOW` statement roundtrip with the same result?  Won't this create the table with explicit indexing instead of implicit `PARTITION ALL` indexing?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58748)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm)

___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit, line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58748#-MQrL-vO1qQ3eNRz3UHk:-MQryLOb4q--4XTcR3ft:b-ue4zzq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b8309bb678afabc85d11022c7a0e8cdbce6f0b9/pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit#L181)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

I might be missing something, but will this `SHOW` statement roundtrip with the same result?  Won't this create the table with explicit indexing instead of implicit `PARTITION ALL` indexing?
</blockquote></details>

I've added a TODO for this, and an issue number.


<!-- Sent from Reviewable.io -->

ajstorm reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58748#-:-MQtr79q7O34nZPG_I7q:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit, line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58748#-MQrL-vO1qQ3eNRz3UHk:-MQtr1M-D0kuJauI2A8q:brcyur7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b8309bb678afabc85d11022c7a0e8cdbce6f0b9/pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit#L181)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

I've added a TODO for this, and an issue number.
</blockquote></details>

Bueno!


<!-- Sent from Reviewable.io -->

otan committed:
sql: implement index changes of PARTITION ALL BY on CREATE TABLE

Release note (sql change): Implemented PARTITION ALL BY syntax for CREATE
TABLE, which automatically partitions the table, and all indexes, with
the same partitioning scheme.

otan commented:
bors r=ajstorm

otan commented:
bors r=ajstorm

otan commented:
bors r-

otan commented:
lol

bors r=ajstorm
